File: api/argus_api/detect.py
import re

from argus_api.models import Kind

_HEADER = re.compile(r"^(from|to|subject|received|return-path|message-id|date|reply-to|authentication-results):", re.I | re.M)
_URL = re.compile(r"^(www\.)?([a-z0-9-]+\.)+[a-z]{2,}(:\d{2,5})?([/?#]\S*)?$", re.I)
_IP_URL = re.compile(r"^(https?://)?\d{1,3}(\.\d{1,3}){3}(:\d{2,5})?([/?#]\S*)?$")
_SCHEME = re.compile(r"^https?://\S+$", re.I)
_PHONE = re.compile(r"^\+?[\d\s\-().]{7,20}$")


def detect_kind(raw: str) -> Kind:
    text = raw.strip()
    if len(_HEADER.findall(text)) >= 2:
        return "email"
    if _PHONE.match(text) and 7 <= sum(ch.isdigit() for ch in text) <= 15:
        return "phone"
    if _SCHEME.match(text) or _IP_URL.match(text) or _URL.match(text):
        return "url"
    return "text"

File: web/src/lib/detect.ts
import type { ScanKind } from "./types";

// Mirrors api/argus_api/detect.py so the page can say what it's reading before anything is sent.
const HEADER = /^(from|to|subject|received|return-path|message-id|date|reply-to|authentication-results):/gim;
const URL = /^(www\.)?([a-z0-9-]+\.)+[a-z]{2,}(:\d{2,5})?([/?#]\S*)?$/i;
const IP_URL = /^(https?:\/\/)?\d{1,3}(\.\d{1,3}){3}(:\d{2,5})?([/?#]\S*)?$/;
const SCHEME = /^https?:\/\/\S+$/i;
const PHONE = /^\+?[\d\s\-().]+$/;

export function detectKind(raw: string): ScanKind | null {
  const text = raw.trim();
  if (!text) return null;
  if ((text.match(HEADER) ?? []).length >= 2) return "email";
  if (IP_URL.test(text)) return "url";
  if (PHONE.test(text) && (text.match(/\d/g) ?? []).length >= 7) return "phone";
  if (SCHEME.test(text) || URL.test(text)) return "url";
  return "text";
}
